### PR TITLE
Add `with edit(layer):` to python

### PR DESCRIPTION
**Pull request by m-kuhn**
Example:

```
with edit(layer):
    f=layer.getFeatures().next()
    f[0]=5
    layer.updateFeature(f)
```

This will automatically call commitChanges() in the end.
If any exception occurs, it will rollBack() all the changes.

### CODE REVIEW

File: python/__init__.py [hunk 1]
```diff
@@ -72,6 +72,27 @@ def __hash__(self):
 except ImportError:
     pass
 
+# Define a `with edit(layer)` statement
```
Comment (NathanW2) [inline python/__init__.py]: Maybe more this to `qgis.core`?

**Comment by NathanW2** [comment]
Nice feature ;)

File: python/__init__.py (hunk 1 above)
Comment (m-kuhn) [inline python/__init__.py]: It is, that's just local and then referenced in core.

Comment (NathanW2) [inline python/__init__.py]: Yeah but there is a `__init__` in the core folder so I think it might be better in there.

Comment (m-kuhn) [inline python/__init__.py]: The NULL as well?

**Comment by m-kuhn** [comment]
I hope this gets more people to edit on the layer instead of the provider.

File: python/__init__.py [hunk 2]
```diff
@@ -72,6 +72,27 @@ def __hash__(self):
...
+    def __init__(self,layer):
+        self.layer = layer
+
+    def __enter__(self):
+        assert self.layer.startEditing()
+        return self.layer
+
+    def __exit__(self, ex_type, ex_value, traceback):
+        if ex_type is None:
+            assert self.layer.commitChanges()
```
Comment (NathanW2) [inline python/__init__.py]: I would also check for the return error and we can return an exception here. `raise QgsFeatureExpection(errors)`.

```
try:
    with edit(layer):
           ....  
except QgsFeatureException:
    print "Doh!"
```

File: python/__init__.py (hunk 1 above)
Comment (NathanW2) [inline python/__init__.py]: Yeah those can move too.  I think we did that before I created the core `__init__` files.

**Comment by nyalldawson** [comment]
nice!!

File: python/__init__.py (hunk 2 above)
Comment (m-kuhn) [inline python/__init__.py]: Wouldn't it be nice to have that defined directly in qgsvectorlayer.sip? This way we get it for free here.

Comment (NathanW2) [inline python/__init__.py]: You could but you don't really gain anything I don't think unless we raise
it from inside the sip bindings, might as well define it in Python to make
it easier.

Comment (m-kuhn) [inline python/__init__.py]: If it's defined in sip we will get the exception even if not using with (like a lot of code currently does). I think that's a big plus.

Comment (NathanW2) [inline python/__init__.py]: Oh I didn't mean define it within the edit class. Just use it there. Make a
new exception class. Or am I missing what you mean.

Comment (m-kuhn) [inline python/__init__.py]: My point is that the edit class re-throws exceptions anyway. And that this exception should not be raised by the edit class but re-raised from one originally raised by QgsVectorLayer::comitChanges().

Comment (m-kuhn) [inline python/__init__.py]: OTOH... That will probably  break tons of existing plugins...

**Pull request merged**